feat(Platform): update actor configuration when receiving control messages from connectors during sync

pedroslopez opened the pull request:
## What

Process connector's config control messages and update the stored configuration accordingly when this kind of message is received.

This PR focuses on making this work for the replication, and a follow-up PR will do the same for the other commands e.g. check, discover)

Here's a[ Loom demo](https://www.loom.com/share/956ef99da8e449f7b5ff5e42478a5957) of it in action.

Closes airbytehq/airbyte#17909
Closes https://github.com/airbytehq/airbyte/issues/17909
Closes https://github.com/airbytehq/airbyte-internal-issues/issues/206

## How

- Create new UpdateConnectorConfigHelper class to call APIs to update configs for a source / destination
- Inject and use the above in DefaultReplicationWorker
- Accept CONTROL messages from the destination / source so we can process them
- In the DefaultReplicationWorker, when we see a config control message call the helper class to store the updated config
- Update the Source/Destination update APIs to handle masking global OAuth parameters

TODO:
- [x] handle initializing the AirbyteApiClient / UpdateConnectorConfigHelper for use in the container-orchestrator in kubernetes environments
- [x] test the above in a real kube env

## Recommended reading order
1. `x.java`
2. `y.python`

## 🚨 User Impact 🚨
No breaking changes - connectors will now be able to update their configs in the middle of a sync.

pedroslopez committed:
track latest config message

pedroslopez committed:
pass new config as part of outputs

pedroslopez committed:
persist new config

pedroslopez committed:
persist config as the messages come through, dont set output

pedroslopez committed:
clean up old implementation

pedroslopez committed:
accept control messages for destinations

pedroslopez committed:
get api client from micronaut

pedroslopez committed:
mask instance-wide oauth params when updating configs

pedroslopez committed:
defaultreplicationworker tests

pedroslopez committed:
formatting

pedroslopez committed:
tests for source/destination handlers

pedroslopez committed:
rm todo

pedroslopez committed:
refactor test a bit to fix pmd

pedroslopez commented on airbyte-commons-worker/src/main/java/io/airbyte/workers/internal/DefaultAirbyteDestination.java:
```diff
@@ -81,8 +82,9 @@ public void start(final WorkerDestinationConfig destinationConfig, final Path jo
     writer = messageWriterFactory.createWriter(new BufferedWriter(new OutputStreamWriter(destinationProcess.getOutputStream(), Charsets.UTF_8)));
 
+    final List<Type> acceptedMessageTypes = List.of(Type.STATE, Type.TRACE, Type.CONTROL);
     messageIterator = streamFactory.create(IOs.newBufferedReader(destinationProcess.getInputStream()))
-        .filter(message -> message.getType() == Type.STATE || message.getType() == Type.TRACE)
+        .filter(message -> acceptedMessageTypes.contains(message.getType()))
         .iterator();
```
Enable reading control messages from destinations

pedroslopez commented on airbyte-commons-worker/src/main/java/io/airbyte/workers/internal/DefaultAirbyteSource.java:
```diff
@@ -86,9 +87,10 @@ public void start(final WorkerSourceConfig sourceConfig, final Path jobRoot) thr
     logInitialStateAsJSON(sourceConfig);
 
+    final List<Type> acceptedMessageTypes = List.of(Type.RECORD, Type.STATE, Type.TRACE, Type.CONTROL);
     messageIterator = streamFactory.create(IOs.newBufferedReader(sourceProcess.getInputStream()))
         .peek(message -> heartbeatMonitor.beat())
-        .filter(message -> message.getType() == Type.RECORD || message.getType() == Type.STATE || message.getType() == Type.TRACE)
+        .filter(message -> acceptedMessageTypes.contains(message.getType()))
```
Enable reading control messages from sources

pedroslopez commented on airbyte-commons-worker/src/test/java/io/airbyte/workers/general/DefaultReplicationWorkerTest.java:
```diff
@@ -231,14 +242,112 @@ void testReplicationRunnableSourceFailure() throws Exception {
+        persistConfigHelper);
+
+    final ReplicationOutput output = worker.run(syncInput, jobRoot);
+    assertEquals(ReplicationStatus.FAILED, output.getReplicationAttemptSummary().getStatus());
```
Note: we will fail the sync if there was an error when persisting the new config.

pedroslopez commented on airbyte-container-orchestrator/src/main/java/io/airbyte/container_orchestrator/ReplicationJobOrchestrator.java:
```diff
@@ -149,6 +153,15 @@ public Optional<String> runJob() throws Exception {
     return Optional.of(Jsons.serialize(replicationOutput));
   }
 
+  private AirbyteApiClient getAirbyteApiClient() {
+    // TODO make this actually functional
+    return new AirbyteApiClient(
+        new ApiClient().setScheme("http")
+            .setHost("airbyte-server-svc")
+            .setPort(8001)
+            .setBasePath("/api"));
+  }
```
This is the main missing piece right now: getting it working and trying it out in a kubernetes environment with the container orchestrator.

pedroslopez commented on airbyte-persistence/job-persistence/src/main/java/io/airbyte/persistence/job/factory/OAuthConfigSupplier.java:
```diff
@@ -85,31 +117,21 @@ public JsonNode injectDestinationOAuthParameters(final UUID destinationDefinitio
+  /**
+   * Gets the OAuth parameter paths as specified in the connector spec and traverses through them
+   */
+  private static void traverseOAuthOutputPaths(final ConnectorSpecification spec,
```
Extracted this logic from the existing injection method so it can be re-used for both injection and masking

pedroslopez commented on airbyte-persistence/job-persistence/src/main/java/io/airbyte/persistence/job/factory/OAuthConfigSupplier.java:
```diff
@@ -47,6 +49,36 @@ public static boolean hasOAuthConfigSpecification(final ConnectorSpecification s
     return spec != null && spec.getAdvancedAuth() != null && spec.getAdvancedAuth().getOauthConfigSpecification() != null;
   }
 
+  public JsonNode maskSourceOAuthParameters(final UUID sourceDefinitionId, final UUID workspaceId, final JsonNode sourceConnectorConfig)
```
Added some new methods to _mask_ oauth parameters from a provided config and ensure we don't write our global oauth params to the connector-specific config / secrets.

pedroslopez commented on airbyte-server/src/main/java/io/airbyte/server/handlers/SourceHandler.java:
```diff
@@ -295,14 +301,14 @@ private void persistSourceConnection(final String name,
                                        final JsonNode configurationJson,
                                        final ConnectorSpecification spec)
       throws JsonValidationException, IOException {
+    final JsonNode oAuthMaskedConfigurationJson = oAuthConfigSupplier.maskSourceOAuthParameters(sourceDefinitionId, workspaceId, configurationJson);
```
New: Source and Destination update APIs now mask global oAuth parameters if they're present. For OSS users without any oauth parameters in their db this will have no effect.

These endpoints were already handling splitting out the secrets to store separately from the config - we're just adding to this and _also_ masking global oauth params.

davinchia replied:
for my understanding, how is this related to persisting configs?

pedroslopez replied:
This is necessary because when we receive a config control message from a connector, it will include those global oauth parameters as part of the config (e.g. our oauth app's client_id + secret). These oauth params are not provided by the user, and are instead injected by the platform. We don't want to persist those as part of the user's secrets when writing back the updated configuration, so we need to mask them again.

alafanechere replied:
Shall we mask these global oAuth parameters on the connector side in the CDK?

pedroslopez replied:
No, the connector doesn't know what's a global auth param and what's provided directly by the user. Note that in the case of OSS users can provide their own oauth client ids and secrets as part of the config, and global auth params are really only enabled in Cloud. I think considering that the control message contains secrets and handling that on the platform side makes the interface easier to work with, and less to have connector devs worry about.

pedroslopez commented on airbyte-commons-worker/src/main/java/io/airbyte/workers/helper/PersistConfigHelper.java:
```diff
@@ -0,0 +1,90 @@
+ * Helper class for workers to persist updates to Source/Destination configs emitted from
+ * AirbyteControlMessages.
+ */
+public class PersistConfigHelper {
```
Extracted the methods to persist config updates for a job into this helper class since there are a few api calls here that we'll very likely re-use when bringing this functionality to the other connector commands.

evantahler replied:
cc @davinchia think we should also have a `PersistStateHelper` and `PersistMetricsAndEstimates` helper like this?

davinchia replied:
Yeap having a various persist helpers sounds generally right.

One thought - what about changing the name to `UpdateConnectorConfigHelper`? Update seems better as it's clear we are persisting to update and not persisting for the first time. We would propagate that name change to the functions here too e.g `updateSource` and `updateDestination`. Can probably drop the `config` suffix since `UpdateConnectorConfigHelper.updateSource` is pretty self-explanatory.

Can we also have a better explanation of why in the javadocs? This describes what and not really the why. We want to make it as easy as possible for a readers to piece the context together without talking to someone.

pedroslopez replied:
I like it! Renamed and added some docs in [bff1d44](https://github.com/airbytehq/airbyte/pull/19811/commits/bff1d44900f3e682c37eb4adc6f9df6db77b2248)

davinchia commented on airbyte-commons-worker/src/main/java/io/airbyte/workers/internal/AirbyteMessageTracker.java:
```diff
@@ -238,12 +238,8 @@ private void handleEmittedOrchestratorMessage(final AirbyteControlMessage contro
-     * persistConfigChange(connectorType, key, configMessage.getConfig().get(key)); // nuance here for
-     * secret storage or not. May need to be async over API for replication orchestrator }
-     */
+    // Currently being persisted as part of the DefaultReplicationWorker
```
note: we shouldn't persist the config in the airbyte message tracker - the tracker is really for stats/metadata tracking and should not have side effects

pedroslopez replied:
Yep, agree! Updated the comment to no longer be a TODO 😛

pedroslopez committed:
fix pmd

pedroslopez committed:
fix test

pedroslopez committed:
add PersistConfigHelperTest

pedroslopez committed:
update message tracker comment

pedroslopez commented:
Still TODO is proper integration with the container-orchestrator. The PR that migrated it to micronaut has just been merged - I'll tackle this soon but want to start getting some eyes on the approach 😄

evantahler commented on airbyte-commons-worker/src/main/java/io/airbyte/workers/general/DefaultReplicationWorker.java:
```diff
@@ -184,7 +190,7 @@ private void replicate(final Path jobRoot,
       // thrown
       final CompletableFuture<?> readFromDstThread = CompletableFuture.runAsync(
-          readFromDstRunnable(destination, cancelled, messageTracker, mdc, timeTracker),
+          readFromDstRunnable(destination, cancelled, messageTracker, persistConfigHelper, mdc, timeTracker, Long.valueOf(jobId)),
```
Reasonable information to have in a job!

davinchia replied:
same thought here on pulling job id into the PersistConfigHelper constructor to simplify the interface

evantahler commented on airbyte-commons-worker/src/main/java/io/airbyte/workers/general/DefaultReplicationWorker.java:
```diff
@@ -250,8 +259,19 @@ private static Runnable readFromDstRunnable(final AirbyteDestination destination
+            LOGGER.info("State in DefaultReplicationWorker from destination: {}", message);
+            messageTracker.acceptFromDestination(message);
+
+            try {
+              if (message.getType() == Type.CONTROL && message.getControl().getType() == AirbyteControlMessage.Type.CONNECTOR_CONFIG) {
+                persistConfigHelper.persistDestinationConfig(jobId, message.getControl().getConnectorConfig().getConfig());
+              }
+            } catch (final ApiException e) {
+              LOGGER.error("Error trying to save updated destination config", e);
+              throw new RuntimeException(e);
+            }
```
👍 that if we can't persist a config update we should fail the sync.
As a note, I read this as a blocking activity - processing updated configs will block all other message processing.  This is probably fine as it won't happen that often in the sync

evantahler commented on airbyte-commons-worker/src/main/java/io/airbyte/workers/helper/PersistConfigHelper.java:
```diff
@@ -0,0 +1,90 @@
+    final SourceApi sourceApi = apiClient.getSourceApi();
+    final SourceRead source = sourceApi.getSource(new SourceIdRequestBody().sourceId(sourceId));
+
+    final SourceRead updatedSource = sourceApi
+        .updateSource(new SourceUpdate()
+            .sourceId(sourceId)
+            .name(source.getName())
+            .connectionConfiguration(Jsons.jsonNode(config.getAdditionalProperties())));
```
I don't see anything in here about authentication... I assume that's handled for us?

pedroslopez replied:
The ApiClientBeanFactory used by micronaut to initialize the ApiClient takes care of setting up the client with any auth-related / evironment-dependent functionality:

https://github.com/airbytehq/airbyte/blob/c30e8f64002d4042121b3b9b6284d75cfc054ea1/airbyte-commons-worker/src/main/java/io/airbyte/workers/config/ApiClientBeanFactory.java#L33

evantahler replied:
☕

evantahler commented on airbyte-commons-worker/src/main/java/io/airbyte/workers/helper/PersistConfigHelper.java:
```diff
@@ -0,0 +1,90 @@
+            .name(source.getName())
+            .connectionConfiguration(Jsons.jsonNode(config.getAdditionalProperties())));
+
+    LOGGER.info("Persisted updated configuration for source {}. New config hash: {}.", sourceId,
+        Hashing.sha256().hashString(updatedSource.getConnectionConfiguration().asText(), StandardCharsets.UTF_8));
```
👍 not logging the actual secrets in the config

evantahler commented on airbyte-commons-worker/src/main/java/io/airbyte/workers/internal/AirbyteMessageTracker.java:
```diff
@@ -238,12 +238,8 @@ private void handleEmittedOrchestratorMessage(final AirbyteControlMessage contro
-     * persistConfigChange(connectorType, key, configMessage.getConfig().get(key)); // nuance here for
-     * secret storage or not. May need to be async over API for replication orchestrator }
-     */
+    // Config updates are being persisted as part of the DefaultNormalizationWorker.
+    // In the future, we could add tracking of these kinds of messages here. Nothing to do for now.
```
That's a great idea! Story added for this: https://github.com/airbytehq/airbyte/issues/19951

evantahler reviewed (commented):
👍 from me! Probably best to wait for someone on the Platform teams to approve

pedroslopez committed:
fix pmd

pedroslopez committed:
Merge branch 'master' into pedroslopez/config-control-msg

pedroslopez committed:
format

pedroslopez committed:
move ApiClientBeanFactory to commons-worker, use in container-orchestrator

pedroslopez commented on airbyte-container-orchestrator/src/main/resources/application.yml:
```diff
@@ -3,4 +3,18 @@ micronaut:
-  config-dir: /config
\ No newline at end of file
+  config-dir: /config
+  control:
+    plane:
+      auth-endpoint: ${CONTROL_PLANE_AUTH_ENDPOINT:}
+  data:
+    plane:
+      service-account:
+        credentials-path: ${DATA_PLANE_SERVICE_ACCOUNT_CREDENTIALS_PATH:}
+        email: ${DATA_PLANE_SERVICE_ACCOUNT_EMAIL:}
+  internal:
+    api:
+      auth-header:
+        name: ${AIRBYTE_API_AUTH_HEADER_NAME:}
+        value: ${AIRBYTE_API_AUTH_HEADER_VALUE:}
+      host: ${INTERNAL_API_HOST}
```
This probably means we need to add more environment variables to the container-orchestrator. I'm in the process of trying this out on a dev cloud environment and will make a PR over there to add the env vars as needed. Haven't been able to fully test due to some incompatibility with the latest OSS branch contents and Cloud that are currently being worked through.

pedroslopez replied:
See https://github.com/airbytehq/airbyte/pull/20724 and https://github.com/airbytehq/airbyte-cloud/pull/3902

davinchia commented on airbyte-commons-worker/src/main/java/io/airbyte/workers/general/DefaultReplicationWorker.java:
```diff
@@ -250,8 +259,19 @@ private static Runnable readFromDstRunnable(final AirbyteDestination destination
+            LOGGER.info("State in DefaultReplicationWorker from destination: {}", message);
+            messageTracker.acceptFromDestination(message);
+
+            try {
+              if (message.getType() == Type.CONTROL && message.getControl().getType() == AirbyteControlMessage.Type.CONNECTOR_CONFIG) {
+                persistConfigHelper.persistDestinationConfig(jobId, message.getControl().getConnectorConfig().getConfig());
+              }
+            } catch (final ApiException e) {
+              LOGGER.error("Error trying to save updated destination config", e);
+              throw new RuntimeException(e);
+            }
+
```
let's this this into private function named along the lines of `updateConnectorConfig` to maintain the same levels of abstraction here.

pedroslopez replied:
👍🏻 done in [51f1976](https://github.com/airbytehq/airbyte/pull/19811/commits/51f19765184523e0ae88d8ffceaaad6e39a772e1)

davinchia commented on airbyte-commons-worker/src/main/java/io/airbyte/workers/general/DefaultReplicationWorker.java:
```diff
@@ -236,8 +243,10 @@ private void replicate(final Path jobRoot,
                                               final Map<String, String> mdc,
-                                              final ThreadedTimeTracker timeHolder) {
+                                              final ThreadedTimeTracker timeHolder,
+                                              final Long jobId) {
```
to avoid adding more parameters to this function, can we make this a part of the PersistConfigHelper constructor?

Why I like this:
- this information is already available when the when the PersistConfigHelper is created,
- the PersistConfigHelper is created per job, so it's ok for it to have job specific information.

pedroslopez replied:
I decided not to add this as part of the helper constructor because I'd like to use this in the other connector commands like check/discover, which don't have a jobId like this. In that follow-up PR I'll pull out some of the logic into a different method for re-use.

davinchia commented on airbyte-commons-worker/build.gradle:
```diff
@@ -10,6 +10,7 @@ dependencies {
     implementation libs.bundles.micronaut
 
     implementation 'io.fabric8:kubernetes-client:5.12.2'
+    implementation 'com.auth0:java-jwt:3.19.2'
```
curious, what is this for?

pedroslopez replied:
This was necessary in order to move the ApiClientBeanFactory into commons-worker (to reuse it in both airbyte-workers and container-orchestrator). It does some token generation that needs this dependency.

pedroslopez replied:
@gosusnp Could you expand a bit on what the issue might be? My change moved the `AirbyteAPiClientBeanFactory` to `airbyte-commons-worker`. I don't see `airbyte-workers` nor `airbyte-commons-worker` being a dependency of `airbyte-cron`.

gosusnp replied:
I was looking for an issue to link for better context but I cannot find it...
I think the problem we had is that micronaut discovers all singleton upon starts, so by having some services in a shared lib led us to start more services than intended. For example, the cron would pick up some beans from the workers and would effectively behave as an extra worker.

@jdpgrailsdev or @benmoriceau might have better context since they were more closely working on this.

davinchia commented on airbyte-commons-worker/src/main/java/io/airbyte/workers/general/DefaultReplicationWorker.java:
```diff
@@ -307,6 +329,15 @@ private static Runnable readFromSrcAndWriteToDstRunnable(final AirbyteSource sou
 
             messageTracker.acceptFromSource(message);
 
+            if (message.getType() == Type.CONTROL && message.getControl().getType() == AirbyteControlMessage.Type.CONNECTOR_CONFIG) {
+              try {
+                persistConfigHelper.persistSourceConfig(jobId, message.getControl().getConnectorConfig().getConfig());
+              } catch (final ApiException e) {
+                LOGGER.error("Error trying to save updated source config", e);
+                throw new RuntimeException(e);
+              }
+            }
+
```
same though here on pulling this out into a private method.

pedroslopez replied:
👍🏻 done in [51f1976](https://github.com/airbytehq/airbyte/pull/19811/commits/51f19765184523e0ae88d8ffceaaad6e39a772e1)

davinchia commented on airbyte-commons-worker/src/main/java/io/airbyte/workers/helper/PersistConfigHelper.java:
```diff
@@ -0,0 +1,90 @@
+    this.apiClient = apiClient;
+  }
+
+  private UUID getConnectionIdFromJobId(final Long jobId) throws ApiException {
```
We've seen some failures in the api calls from the data plane to the control plane and are still trying to figure out why.

In the mean time, I've added a [retry wrapper in this PR](https://github.com/airbytehq/airbyte/pull/19693/files#diff-81b9db1a46774bfa5dc46a3800a820b7c6596c799822e2012761fb37d65c9eb5R147). We probably want to use this wrapper for all the api calls in this class.

pedroslopez replied:
👍🏻 done in [8203432](https://github.com/airbytehq/airbyte/pull/19811/commits/8203432df13dd43adf060dc7704ce68de0da27ef)

davinchia reviewed (commented):
Thanks for the tag @pedroslopez . This looks reasonable.

I left some suggestions. I might not be able to sneak time to take another look during PTO so please feel free to merge in after thinking about these suggestions. We can always revisit in later PR.

I focused my review on the platform classes. I did not scrutinise the tests or the oauth pieces.

I would also like someone from the workflows team to take a look. @gosusnp tagging you since you've expressed interest in side effects during replication. Please tag someone from workflows in if you feel there is a better reviewer.

pedroslopez committed:
pull out config updating to separate methods

pedroslopez committed:
add jitter

pedroslopez committed:
rename PersistConfigHelper -> UpdateConnectorConfigHelper, docs

pedroslopez committed:
Merge branch 'master' into pedroslopez/config-control-msg

pedroslopez committed:
fix exception type

pedroslopez committed:
fmt

alafanechere commented on airbyte-container-orchestrator/src/main/resources/application.yml:
```diff
@@ -3,4 +3,18 @@ micronaut:
-  config-dir: /config
\ No newline at end of file
+  config-dir: /config
+  control:
+    plane:
+      auth-endpoint: ${CONTROL_PLANE_AUTH_ENDPOINT:}
+  data:
+    plane:
+      service-account:
+        credentials-path: ${DATA_PLANE_SERVICE_ACCOUNT_CREDENTIALS_PATH:}
+        email: ${DATA_PLANE_SERVICE_ACCOUNT_EMAIL:}
+  internal:
+    api:
+      auth-header:
+        name: ${AIRBYTE_API_AUTH_HEADER_NAME:}
+        value: ${AIRBYTE_API_AUTH_HEADER_VALUE:}
+      host: ${INTERNAL_API_HOST}
```
Could you explain the purpose of this change?

pedroslopez replied:
This is to be able to initialize and use AirbyteApiClient from the container orchestrator. It's a micronaut configuration file that is used when configuring the API client. An equivalent configuration is present in airbyte-workers.

alafanechere reviewed (approved):
Nice work! Loved the demo.

pedroslopez commented on airbyte-commons-worker/src/main/java/io/airbyte/workers/helper/UpdateConnectorConfigHelper.java:
```diff
@@ -0,0 +1,115 @@
+   * Updates the Source from a sync job ID with the provided Configuration. Secrets and OAuth
+   * parameters will be masked when saving.
+   */
+  public void updateSource(final Long jobId, final Config config) {
+    final UUID connectionId = getConnectionIdFromJobId(jobId);
+
+    final ConnectionRead connection = AirbyteApiClient.retryWithJitter(
+        () -> apiClient.getConnectionApi().getConnection(new ConnectionIdRequestBody().connectionId(connectionId)),
+        "get connection");
+    final UUID sourceId = connection.getSourceId();
+
+    final SourceApi sourceApi = apiClient.getSourceApi();
+    final SourceRead source = AirbyteApiClient.retryWithJitter(
+        () -> sourceApi.getSource(new SourceIdRequestBody().sourceId(sourceId)),
+        "get source");
+
+    final SourceRead updatedSource = AirbyteApiClient.retryWithJitter(
+        () -> sourceApi
+            .updateSource(new SourceUpdate()
+                .sourceId(sourceId)
+                .name(source.getName())
+                .connectionConfiguration(Jsons.jsonNode(config.getAdditionalProperties()))),
+        "update source");
+
+    LOGGER.info("Persisted updated configuration for source {}. New config hash: {}.", sourceId,
+        Hashing.sha256().hashString(updatedSource.getConnectionConfiguration().asText(), StandardCharsets.UTF_8));
+
+  }
+
+  /**
+   * Updates the Destination from a sync job ID with the provided Configuration. Secrets and OAuth
+   * parameters will be masked when saving.
+   */
+  public void updateDestination(final Long jobId, final Config config) {
```
Yeah, it's tricky because even though Sources/Destinations are equivalent in the DB, they're handled  as completely different in the java classes. Unfortunately there's tons of this kind of duplication in the java codebase for handling sources and destinations. All the methods for interacting with the classes are also type-specific, e.g. `.sourceId`, `.updateSource` etc.

gosusnp commented on airbyte-commons-worker/src/main/java/io/airbyte/workers/helper/UpdateConnectorConfigHelper.java:
```diff
@@ -0,0 +1,115 @@
+   * parameters will be masked when saving.
+   */
+  public void updateSource(final Long jobId, final Config config) {
+    final UUID connectionId = getConnectionIdFromJobId(jobId);
```
This part feels very inefficient, taking a step back, when we setup the jobs, we have all the pieces, `connectionId`, `sourceId`, `destinationId`, *but* since we only pass `jobId`, we need to have an extra API call to retrieve the sourceId/destinationId.
Can we pass those instead of just the jobId and remove the need of the `connectionApi`.

pedroslopez replied:
Agree! I've moved to specifying the sourceId and destinationId as part of the StandardSyncInput and passing that along so we have the information available. This means we can call get/update directly using these IDs rather than have to look up the job, then the connection, then the source/destination.

gosusnp commented on airbyte-commons-worker/src/main/java/io/airbyte/workers/helper/UpdateConnectorConfigHelper.java:
```diff
@@ -0,0 +1,115 @@
+
+  private final AirbyteApiClient apiClient;
+
+  public UpdateConnectorConfigHelper(final AirbyteApiClient apiClient) {
```
We're moving towards a more fine grained injection, can we injection `sourceApi` and `destinationApi` rather than the `AirbyteApiClient`? (I have a comment down that aims at removing the dependency on `connectionApi`

pedroslopez replied:
Done in https://github.com/airbytehq/airbyte/pull/19811/commits/6ecac44ac2f7c0d1e2ea4be33cb5079fce913b00

gosusnp commented on airbyte-commons-worker/src/main/java/io/airbyte/workers/helper/UpdateConnectorConfigHelper.java:
```diff
@@ -0,0 +1,115 @@
+ * useful for migrating configuration to a new version or for enabling connectors that require
+ * single-use or short-lived OAuth tokens.
+ */
+public class UpdateConnectorConfigHelper {
```
nit: is this just a `ConnectorConfigUpdater` rather than Helper?

pedroslopez replied:
Renamed in https://github.com/airbytehq/airbyte/pull/19811/commits/c20f61e673363d293d33ea3ccf7b5c6fb38c0fea

gosusnp reviewed (commented):
Change overall looks good, most of my comments are around the config update piece.

pedroslopez committed:
move message type check into runnable

pedroslopez committed:
formatting

pedroslopez committed:
pass api client env vars to container orchestrator

pedroslopez committed:
pass micronaut envs to container orchestrator

pedroslopez committed:
print stacktrace for debugging

pedroslopez committed:
Merge branch 'master' into pedroslopez/config-control-msg

pedroslopez committed:
different api host for container orchestrator

pedroslopez committed:
fix default env var

pedroslopez committed:
Merge branch 'master' into pedroslopez/config-control-msg

pedroslopez committed:
format

pedroslopez committed:
fix errors after merge

pedroslopez committed:
Merge branch 'master' into pedroslopez/config-control-msg

pedroslopez committed:
set source and destination actor id as part of the sync input

pedroslopez committed:
fix: get destination definition

pedroslopez committed:
fix null ptr

pedroslopez committed:
remove "actor" from naming

pedroslopez committed:
fix missing change from rename

pedroslopez committed:
revert ContainerOrchestratorConfigBeanFactory changes

pedroslopez committed:
Merge branch 'master' into pedroslopez/config-control-msg

pedroslopez committed:
Merge branch 'master' into pedroslopez/config-control-msg

pedroslopez committed:
inject sourceapi/destinationapi directly rather than airbyteapiclient

pedroslopez committed:
UpdateConnectorConfigHelper -> ConnectorConfigUpdater

pedroslopez commented on airbyte-commons-worker/src/main/java/io/airbyte/workers/config/ApiClientBeanFactory.java:
```diff
@@ -67,6 +68,11 @@ public SourceApi sourceApi(final ApiClient apiClient) {
     return new SourceApi(apiClient);
   }
 
+  @Singleton
+  public DestinationApi destinationApi(final ApiClient apiClient) {
+    return new DestinationApi(apiClient);
+  }
```
Added so we have the singleton available for injection directly - followed what was already being done for `SourceApi`

pedroslopez commented on airbyte-config/config-models/src/main/resources/types/StandardSyncInput.yaml:
```diff
@@ -19,6 +21,14 @@ properties:
   prefix:
     description: Prefix that will be prepended to the name of each stream when it is written to the destination.
     type: string
+  sourceId:
+    description: Actor ID for the source used in the sync - this is used to update the actor configuration when requested.
+    type: string
+    format: uuid
+  destinationId:
+    description: Actor ID for the destination used in the sync - this is used to update the actor configuration when requested.
+    type: string
+    format: uuid
```
sync inputs now have sourceId and destinationId so we can refer to them directly when updating configs

pedroslopez commented on airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/activities/GenerateInputActivityImpl.java:
```diff
@@ -90,25 +90,18 @@ public GeneratedJobInput getSyncWorkflowInput(final SyncInput input) {
-          .filter(standardDestinationDefinition -> Objects.nonNull(standardDestinationDefinition.getNormalizationConfig()))
-          .map(standardDestinationDefinition -> standardDestinationDefinition.getNormalizationConfig().getNormalizationIntegrationType())
-          .orElse(null);
+      final UUID connectionId = UUID.fromString(job.getScope());
+      final StandardSync standardSync = configRepository.getStandardSync(connectionId);
+
+      final StandardDestinationDefinition destinationDefinition =
+          configRepository.getDestinationDefinitionFromDestination(standardSync.getDestinationId());
+      final String destinationNormalizationDockerImage = destinationDefinition.getNormalizationConfig() != null
+          ? DockerUtils.getTaggedImageName(destinationDefinition.getNormalizationConfig().getNormalizationRepository(),
+              destinationDefinition.getNormalizationConfig().getNormalizationTag())
+          : null;
+      final String normalizationIntegrationType =
+          destinationDefinition.getNormalizationConfig() != null ? destinationDefinition.getNormalizationConfig().getNormalizationIntegrationType()
+              : null;
```
While I was in here, I refactored how we were looking up the destination definition from the sync. We need to get the StandardSync anyway to retrieve the sourceId and destinationId for config updating now, so we can simplify to get the definition from the destinationId directly. 

This used to be looking up all destination definitions and filtering based on docker image + tag combo, which also seemed a bit error prone if we decide to move into a world where multiple definitions could have the same docker image (which I think _will_ be the case)

The functionality should be equivalent.

pedroslopez commented on airbyte-commons-worker/src/main/java/io/airbyte/workers/helper/UpdateConnectorConfigHelper.java:
```diff
@@ -0,0 +1,115 @@
+  public void updateSource(final Long jobId, final Config config) {
+    final UUID connectionId = getConnectionIdFromJobId(jobId);
+
+    final ConnectionRead connection = AirbyteApiClient.retryWithJitter(
```
Discussed offline - we will be blocking for now and revisit later since control messages should not be a common thing during the sync (_maybe_ 1 per hour depending on the connector)

pedroslopez commented on airbyte-commons-worker/src/main/java/io/airbyte/workers/general/DefaultReplicationWorker.java:
```diff
@@ -306,6 +319,7 @@ private static Runnable readFromSrcAndWriteToDstRunnable(final AirbyteSource sou
             final AirbyteMessage message = mapper.mapMessage(airbyteMessage);
 
             messageTracker.acceptFromSource(message);
+            acceptAndUpdateSrcConfig(jobId, message, updateConnectorConfigHelper);
```
Updated to `acceptSrcControlMessage` and `acceptDstControlMessage` methods so we know what's. being passed through

pedroslopez commented:
@gosusnp There's a separate piece in https://github.com/airbytehq/airbyte/pull/20724 that will be merged into this branch to set things up properly for the container orchestrator in kube environments, but I've addressed your comments and would love another look!

Notably the sourceId and destinationId are now being passed as part of the StandardSyncInput.

pedroslopez committed:
rm log

pedroslopez committed:
fix test

gosusnp commented on airbyte-commons-worker/src/main/java/io/airbyte/workers/general/DefaultReplicationWorker.java:
```diff
@@ -267,8 +276,14 @@ private static Runnable readFromDstRunnable(final AirbyteDestination destination
+            messageTracker.acceptFromDestination(message);
+
+            if (message.getType() == Type.CONTROL) {
+              acceptDstControlMessage(destinationConfig, message.getControl(), connectorConfigUpdater);
```
We may want to have a `try...catch` here, otherwise, if we fail to update the config, we'd end up failing the sync.
Same comment for the source.

pedroslopez replied:
Originally I had intentionally made syncs fail if we couldn't update the config, but after thinking about this some more I agree we shouldn't fail because:

- Syncing data is our top priority and what we care about. We should keep moving data.
- These messages would be emitted for two main reasons at the moment: config migrations and handling short-lived or one-time-use tokens for oauth. 
    - For config migrations, we should be able to try again on the next sync. 
    - For the OAuth case, we could see another CONTROL message later on that might get correctly updated, or if it's really an issue we'll get an error on the next sync anyway (so we should let the current sync run its course).

The benefit of failing is we know right away something didn't go quite right, and we would want to know about these errors. However it might be better surfaced as some kind of notice rather than a full failure.

Updated to just log the error in [e7f4736](https://github.com/airbytehq/airbyte/pull/19811/commits/e7f4736e343d0ccc2fc1b8716eda30588a462fa8)

gosusnp reviewed (commented):
Looks good, one last question about how we want to handle config update errors.

pedroslopez committed:
dont fail on config update error

davinchia commented on airbyte-commons-worker/src/main/java/io/airbyte/workers/general/DefaultReplicationWorker.java:
```diff
@@ -391,6 +420,22 @@ private static Runnable readFromSrcAndWriteToDstRunnable(final AirbyteSource sou
+  private static void acceptSrcControlMessage(final WorkerSourceConfig sourceConfig,
+                                              final AirbyteControlMessage controlMessage,
+                                              final ConnectorConfigUpdater connectorConfigUpdater) {
+    if (controlMessage.getType() == AirbyteControlMessage.Type.CONNECTOR_CONFIG) {
```
are there other kinds of control messages today?

pedroslopez replied:
Nope, this is the only one for now!

davinchia commented on airbyte-commons-worker/src/main/java/io/airbyte/workers/general/DefaultReplicationWorker.java:
```diff
@@ -391,6 +420,22 @@ private static Runnable readFromSrcAndWriteToDstRunnable(final AirbyteSource sou
     };
   }
 
+  private static void acceptSrcControlMessage(final WorkerSourceConfig sourceConfig,
```
we can simplify this method signature by passing in the id instead of the entire config

davinchia replied:
this simplification also extends to the various runnable method interfaces e.g ` readFromDstRunnable` doesn't need the `destinationConfig`. Improves readability since it's clearer what scope requires what data.

pedroslopez replied:
👍🏻 updated in [d6eb05f](https://github.com/airbytehq/airbyte/pull/19811/commits/d6eb05f4a9a4edfee6c566ed888a7711f9c41449)

davinchia commented on airbyte-commons-worker/src/main/java/io/airbyte/workers/helper/ConnectorConfigUpdater.java:
```diff
@@ -0,0 +1,87 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
```
+1 for the javadocs!

davinchia reviewed (approved):
Thanks Pedro!

Some minor comments on the code readability in the DefaultReplicationWorker class. Otherwise looks good!

pedroslopez committed:
pass id, not full config to runnables/accept control message

pedroslopez committed:
Merge branch 'master' into pedroslopez/config-control-msg

pedroslopez committed:
add new config required for api client

pedroslopez committed:
add test file

pedroslopez committed:
Merge branch 'master' into pedroslopez/config-control-msg

pedroslopez committed:
Merge branch 'master' into pedroslopez/config-control-msg

pedroslopez committed:
fix test compatibility

pedroslopez committed:
Merge branch 'master' into pedroslopez/config-control-msg

pedroslopez committed:
mount data plane credentials secret to container orchestrator (#20724)

* mount data plane credentials secret to container orchestrator

* rm copy-pasta

* properly handle empty strings

* set env vars like before

* use the right config vars
